[#58] Update device status

felixlapierre opened the pull request:
# General info
<!--Delete or put N/A for any irrelevant sections-->

**Issue number**: #58 

Note that this PR contains the changes in #165 as well, you can ignore those when reviewing.

**Task description**: 

 - Added two endpoints to the backend: `/decoder/{serial}/streams` and `/encoder/{serial}/streams`. These endpoints update the time of last communication of the encoder/decoder to the current instant, and return a list with all the streams that involve that device.
 - In the frontend, when getting the device information, the status of each device is no longer hardcoded to `online`. If the device's last communication is within the last ten minutes, it sets it to online. If it has been more than 10 minutes, the device status is marked offline. If the value of `lastCommunication` is null, it is set to `pending`. This would happen if the device has not retrieved its streams yet.
- Also did a small fix in the frontend, since the names of the inputs/outputs properties of decoders/encoders changed. Also added a protection so the page doesn't crash if a device has no channels.
 
![image](https://user-images.githubusercontent.com/32281260/100280824-cc76a880-2f36-11eb-9394-d3ef6f0dd3ae.png)

# Testing

**Test coverage**:
Tests coming soon to this PR!

**To do before merging**:
- [x] Add unit tests to backend
- [x] Add unit tests to frontend

FGRCL commented on src/main/java/org/beanpod/switchboard/controller/DecoderController.java:
```diff
@@ -67,10 +73,17 @@
+
+  @GetMapping("/{serialNumber}/streams")
+  public ResponseEntity<List<StreamModel>> getDecoderStreams(@PathVariable String serialNumber) {
+    List<StreamDTO> streamDtos = decoderService.getDecoderStreams(serialNumber);
+    List<StreamModel> streamModelList = streamMapper.toModel(streamDtos);
+    return ResponseEntity.ok(streamModelList);
```
We should wrap this is a catch or use `Optional` to throw a general error if anything wrong happens.

felixlapierre replied:
The CustomRestExceptionHandler should catch any other errors, but i'll wrap it in Optional to handle any null return values

FGRCL commented on src/main/java/org/beanpod/switchboard/controller/EncoderController.java:
```diff
@@ -69,10 +69,16 @@
+
+  @GetMapping("/{serialNumber}/streams")
+  public ResponseEntity<List<StreamModel>> getEncoderStreams(@PathVariable String serialNumber) {
+    List<StreamDTO> streamDtos = encoderService.getEncoderStreams(serialNumber);
+    return ResponseEntity.ok(streamMapper.toModel(streamDtos));
```
Would be good to wrap this too

FGRCL commented on src/main/java/org/beanpod/switchboard/dto/mapper/StreamMapper.java:
```diff
@@ -15,5 +17,9 @@
 
   StreamDTO toDto(StreamEntity streamEntity);
 
+  List<StreamDTO> toDto(List<StreamEntity> streamEntityList);
```
Can rename this  to `toDtos` since it's a list?

felixlapierre replied:
Renamed it to `toDtoList`

FGRCL commented on src/main/java/org/beanpod/switchboard/dto/mapper/StreamMapper.java:
```diff
@@ -15,5 +17,9 @@
+
   StreamModel toModel(StreamDTO streamDto);
+
+  List<StreamModel> toModel(List<StreamDTO> streamDtoList);
```
Can rename this  to `toModels` since it's a list?

felixlapierre replied:
Renamed it to `toModelList`

FGRCL commented on src/main/java/org/beanpod/switchboard/repository/StreamRepository.java:
```diff
@@ -19,4 +20,10 @@
       @Param(value = "inputChannelId") long inputChannelId,
       @Param(value = "outputChannelId") long outputChannelId);
+
+  @Query("SELECT s FROM Stream s WHERE s.outputChannel.encoder.serialNumber = :encoderSerial")
+  List<StreamEntity> getEncoderStreams(@Param(value = "encoderSerial") String encoderSerial);
+
+  @Query("SELECT s FROM Stream s WHERE s.inputChannel.decoder.serialNumber = :decoderSerial")
```
Short and sweet, I like it.

ButlerJAL replied:
We possibly want to investigate the JPA method name query features to avoid writing SQL within our Java code. This comment applies for all other instances of writing SQL queries within our Java code base.

FGRCL commented on src/main/java/org/beanpod/switchboard/service/DecoderService.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public interface DecoderService {
+  List<StreamDTO> getDecoderStreams(String decoderSerialNumber);
+}
```
New layer, I like it!

felixlapierre replied:
backend is like an onion
many layers
make you cry

FGRCL commented on src/main/java/org/beanpod/switchboard/service/DecoderServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Override
+  public List<StreamDTO> getDecoderStreams(String decoderSerialNumber) {
+    Optional<DecoderDTO> decoderOptional = decoderDao.findDecoder(decoderSerialNumber);
+    if (decoderOptional.isEmpty()) {
+      throw new ExceptionType.DeviceNotFoundException(decoderSerialNumber);
+    }
+    DecoderDTO decoder = decoderOptional.get();
```
I believe you can replace this with a single `orElseThrow()`

FGRCL commented on src/main/java/org/beanpod/switchboard/service/EncoderServiceImpl.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Override
+  public List<StreamDTO> getEncoderStreams(String encoderSerialNumber) {
+    Optional<EncoderDTO> encoderOptional = encoderDao.findEncoder(encoderSerialNumber);
+    if (encoderOptional.isEmpty()) {
+      throw new ExceptionType.DeviceNotFoundException(encoderSerialNumber);
+    }
+    EncoderDTO encoderDto = encoderOptional.get();
```
Same thing, can be replace by a `orElseThrow`.

FGRCL commented on src/main/java/org/beanpod/switchboard/util/DateUtil.java:
```diff
@@ -0,0 +1,17 @@
+
+/**
+ * Utility class that returns the date of the current time in UTC time zone Useful for injecting
+ * test values for current time in tests.
```
that comment says it's for tests but you use it in a service.

felixlapierre replied:
I removed that part of the comment. I put it because this class only exists so i can do dependency injection in tests, but it's probably more misleading than anything else.

FGRCL commented on src/test/java/org/beanpod/switchboard/controller/EncoderControllerTest.java:
```diff
@@ -142,12 +148,28 @@ final void testUpdateEncoder() {
+  final void testGetEncoderStreams() {
+    when(encoderService.getEncoderStreams(any(String.class)))
+        .thenReturn(List.of(StreamFixture.getStreamDto()));
+    when(streamMapper.toModel(anyList())).thenReturn(List.of(StreamFixture.getStreamModel()));
```
we should add a fixture method to return a list of streams

FGRCL commented on src/test/java/org/beanpod/switchboard/controller/EncoderControllerTest.java:
```diff
@@ -142,12 +148,28 @@ final void testUpdateEncoder() {
+
+  @Test
+  final void testGetEncoderStreams() {
+    when(encoderService.getEncoderStreams(any(String.class)))
+        .thenReturn(List.of(StreamFixture.getStreamDto()));
+    when(streamMapper.toModel(anyList())).thenReturn(List.of(StreamFixture.getStreamModel()));
```
I like to use the explicit call values when mocking instead of `any` because it helps catch more errors when changes are made

felixlapierre replied:
The reason I did this instead of putting explicit call values is that if the value changes, the method will return null instead and the error message becomes less evident.

Like this, if you call with the wrong thing, then it will fail at the stage of the `verify` with a nice little diff telling you what properties don't match.

FGRCL replied:
That's fair

FGRCL reviewed (changes requested):
Only reviewed the backend for now, I have several suggestions.

ButlerJAL commented on frontend/src/api/DeviceApi.js:
```diff
@@ -4,28 +4,49 @@ import * as SampleData from "./SampleData";
+            getStatus(sender.lastCommunication),
             channels,
-            ["Sample sender"]
+            ["Additional Device details go here"]
```
Is it appropriate to hard code such an extras field when returning new DeviceInfo on a HTTP GET request?

ButlerJAL commented on frontend/src/api/DeviceApi.js:
```diff
@@ -41,17 +62,29 @@ export function getReceivers(callback) {
+            getStatus(receiver.lastCommunication),
             channels,
-            ["Sample Reciever"]
+            ["Additional Device details go here"]
```
Same as previous review comment

felixlapierre commented:
Note that I renamed one of the `__tests__` folders to `tests`. The reason I did this is that I added `DeviceFixture.js` for creating test data objects. The test runner was trying to run my fixture class as a file containing tests, because it tries to run every .js file in a `__tests__` folder.

The fix was to rename the folder to `tests` since then it only runs files that end in `.test.js`. Had to adjust pom.xml to inform the code coverage checker aswell.

MrJCipherButtles reviewed (dismissed):
I approve these changes. I left a few comments pointing to things that we might want to investigate in a future task. Overall, I like the better separation between our data and service layers on the code you worked on. I see a few fixes were done end to end to update device status. The test coverage on the new code is good. Although there are more issues to solve regarding us reworking our implementation, these changes are a step in the right direction.

MrJCipherButtles reviewed (approved):
reapprove

FGRCL reviewed (approved):
My comments were adressed
